lib/compiler/src/sys_core_fold.erl

```diff
@@ -1444,40 +1445,34 @@ sub_del_var(#c_var{name=V}, #sub{v=S,s=Scope,t=Tdb}=Sub) ->
 	false ->
 	    Sub#sub{s=cerl_sets:add_element(V, Scope)};
 	true ->
-	    Sub#sub{v=orddict:erase(V, S),t=kill_types(V, Tdb)}
+	    Sub#sub{v=maps:remove(V, S),t=kill_types(V, Tdb)}
     end.
 
 sub_subst_var(#c_var{name=V}, Val, #sub{v=S0}) ->
     %% Fold chained substitutions.
-    [{V,Val}] ++ [ {K,Val} || {K,#c_var{name=V1}} <- S0, V1 =:= V].
+    [{V,Val}] ++ [ {K,Val} || {K,#c_var{name=V1}} <- maps:to_list(S0), V1 =:= V].
 
 sub_add_scope(Vs, #sub{s=Scope0}=Sub) ->
     Scope = foldl(fun(V, S) when is_integer(V); is_atom(V) ->
 			  cerl_sets:add_element(V, S)
 		  end, Scope0, Vs),
     Sub#sub{s=Scope}.
 
-sub_subst_scope(#sub{v=S0,s=Scope}=Sub) ->
-    Initial = case S0 of
-                  [{NegInt,_}|_] when is_integer(NegInt), NegInt < 0 ->
-                      NegInt - 1;
-                  _ ->
-                      -1
-              end,
-    S = sub_subst_scope_1(cerl_sets:to_list(Scope), Initial, S0),
-    Sub#sub{v=orddict:from_list(S)}.
-
-%% The keys in an orddict must be unique. Make them so!
+sub_subst_scope(#sub{v=V0,s=Scope,c=C0}=Sub) ->
+    {C,V} = sub_subst_scope_1(cerl_sets:to_list(Scope), C0, V0),
+    Sub#sub{v=V,c=C}.
+
+%% The keys in an maps must be unique. Make them so!
 sub_subst_scope_1([H|T], Key, Acc) ->
-    sub_subst_scope_1(T, Key-1, [{Key,#c_var{name=H}}|Acc]);
-sub_subst_scope_1([], _, Acc) -> Acc.
+    sub_subst_scope_1(T, Key-1, maps:put(Key, #c_var{name=H}, Acc));
+sub_subst_scope_1([], Key, Acc) -> {Key,Acc}.
 
 sub_is_val(#c_var{name=V}, #sub{v=S,s=Scope}) ->
     %% When the bottleneck in sub_del_var/2 was eliminated, this
     %% became the new bottleneck. Since the scope contains all
     %% live variables, a variable V can only be the target for
     %% a substitution if it is in the scope.
-    cerl_sets:is_element(V, Scope) andalso v_is_value(V, S).
+    cerl_sets:is_element(V, Scope) andalso v_is_value(V, maps:to_list(S)).
```

Comment:
@bjorng Yes, this is the part that worries me too. I did benchmark two or three big modules in Elixir and I also could not measure anything relevant. Profiling showed improvements but they did not translate to actual improvements - as above.

The best would be if we could rewrite part of this code to rely on map lookups but at the moment I don't have enough knowledge of this code to devise alternatives.

Maybe the best course of action for now is to close this pull request. I can revisit it later once I am more familiar with sys_core_fold.